staticd/static_bfd.c

```diff
@@ -0,0 +1,504 @@
+	onlink = yang_dnode_exists(dnode, "../onlink")
+		 && yang_dnode_get_bool(dnode, "../onlink");
+	mhop = yang_dnode_exists(dnode, "./multi-hop")
+		&& yang_dnode_get_bool(dnode, "./multi-hop");
+
+	switch (sn->type) {
+	case STATIC_IPV4_GATEWAY_IFNAME:
+	case STATIC_IPV6_GATEWAY_IFNAME:
+		use_interface = true;
+		/* FALLTHROUGH */
+	case STATIC_IPV4_GATEWAY:
+	case STATIC_IPV6_GATEWAY:
+		if (sn->type == STATIC_IPV4_GATEWAY
+		    || sn->type == STATIC_IPV4_GATEWAY_IFNAME)
+			family = AF_INET;
+		else
+			family = AF_INET6;
+		break;
+
+	case STATIC_IFNAME:
+	case STATIC_BLACKHOLE:
+	default:
+		zlog_err("%s: invalid next hop type", __func__);
+		return;
+	}
+
+	/* Reconfigure or allocate new memory. */
+	if (sn->bsp == NULL)
+		sn->bsp = bfd_sess_new(static_next_hop_bfd_updatecb, sn);
+
+	if (sn->nh_vrf_id != VRF_DEFAULT)
+		bfd_sess_set_vrf(sn->bsp, sn->nh_vrf_id);
+
+	if (family == AF_INET)
+		bfd_sess_set_ipv4_addrs(sn->bsp, NULL, &sn->addr.ipv4);
+	else
+		bfd_sess_set_ipv6_addrs(sn->bsp, NULL, &sn->addr.ipv6);
+
+	bfd_sess_set_interface(sn->bsp, use_interface ? sn->ifname : NULL);
+
+	bfd_sess_set_profile(sn->bsp, use_profile ? yang_dnode_get_string(
+					      dnode, "./profile")
+						  : NULL);
+
+	if (onlink || mhop == false)
+		bfd_sess_set_hop_count(sn->bsp, BFD_SINGLE_HOP_COUNT);
+	else
+		bfd_sess_set_hop_count(sn->bsp, BFD_MULTI_HOP_MAX_HOP_COUNT);
+
+	/* Install or update the session. */
+	bfd_sess_install(sn->bsp);
+
+	/* Update current path status. */
+	sn->path_down = (bfd_sess_status(sn->bsp) != BSS_UP);
+}
+
+void static_next_hop_bfd_monitor_disable(struct static_nexthop *sn)
+{
+	bfd_sess_free(&sn->bsp);
+
+	/* Reset path status. */
+	sn->path_down = false;
+}
+
+void static_next_hop_bfd_multi_hop(struct static_nexthop *sn, bool mhop)
+{
+	if (sn->bsp == NULL)
+		return;
+
+	bfd_sess_set_hop_count(sn->bsp, mhop ? BFD_MULTI_HOP_MAX_HOP_COUNT
+			       : BFD_SINGLE_HOP_COUNT);
+	bfd_sess_install(sn->bsp);
+}
+
+void static_next_hop_bfd_profile(struct static_nexthop *sn, const char *name)
+{
+	if (sn->bsp == NULL)
+		return;
+
+	bfd_sess_set_profile(sn->bsp, name);
+	bfd_sess_install(sn->bsp);
+}
+
+/*
+ * Route group BFD monitoring settings.
+ */
+static struct static_group_member *
+static_group_member_new(struct static_route_group *srg,
+			struct static_nexthop *sn)
+{
+	struct static_group_member *sgm;
+
+	/* Remove next hop specific BFD configuration if any. */
+	bfd_sess_free(&sn->bsp);
+
+	sgm = XCALLOC(MTYPE_ROUTE_GROUP_MEMBER, sizeof(*sgm));
+	sgm->sgm_sn = sn;
+	TAILQ_INSERT_TAIL(&srg->srg_sgmlist, sgm, sgm_entry);
+
+	return sgm;
+}
+
+static void static_group_member_free(struct static_group_member **sgm)
+{
+	struct static_route_group *srg;
+	struct static_nexthop *sn;
+
+	if ((*sgm) == NULL)
+		return;
+
+	srg = (*sgm)->sgm_srg;
+
+	/* Set default next hop installation status. */
+	sn = (*sgm)->sgm_sn;
+	if (sn->path_down) {
+		sn->path_down = false;
+		static_zebra_route_add(sn->pn, true);
```

Comment:
add true ?

Reply:
after resetting the bfd group from this path, if bfd was down, retry to add the routes after resetting path_down.
the call is appropriate, here